File: apps/desktop-macosx/WallpaperInfoApp/WallpaperInfoApp/WallpaperService.swift
//
//  WallpaperService.swift
//  WallpaperInfoApp
//
//  Created by Brad Park on 4/11/20.
//  Copyright © 2020 Brad Park. All rights reserved.
//

import Foundation
import Cocoa
import NIOPosix

class WallpaperService {
    
    init() {
        activateHotkey()
        let bonjourService = BPBonjourService()
        bonjourService.start()
        let fileTransferServer = BPFileReceiver()
        fileTransferServer.start()
    }
    
    deinit {
        if (_started) {
            onDestroy()
        }
    }
    
    func startService() {
        onStartCommand()
    }

    func stopService() {
        _imageFileManager.stopWatching()
        onDestroy()
    }

    private func onDestroy() {
        // stop timer when service is destroyed to prevent service is rerunning
        _platformTimer?.pause()

        // update widget for default view
        let wpsi = WallpaperServiceInfo.getInstance()
        wpsi.setcurrentWPath(nil)
        wpsi.setThumbnail(nil)
        
        if (_started) {
			// removing observer
            NotificationCenter.default.removeObserver(self)
            _started = false
        }
        broadcastServiceUpdate()
    }
	
	public func broadcastReceiver(_ action:Int, _ extras:Any?) {
        if (!_started) {
            return
        }
		switch action {
		case MSG.SET_THEME:
            BPUtil.BPLog("WallpaperServicebroadcastReceiver - MSG.SET_THEME")
			if let theme = Theme(rawValue: extras as! Int) {
				updateServiceTheme(theme: theme)
				broadcastServiceUpdate()
			}
		case MSG.PREVIOUS:
            BPUtil.BPLog("WallpaperServicebroadcastReceiver - MSG.PREVIOUS")
			restartTimer()
			naviageWallpaper(offset: -1)
		case MSG.NEXT:
            BPUtil.BPLog("WallpaperServicebroadcastReceiver - MSG.NEXT")
			restartTimer()
			naviageWallpaper(offset: 1)
		case MSG.PREVIOUS_THEME:
            BPUtil.BPLog("WallpaperServicebroadcastReceiver - MSG.PREVIOUS_THEME")
			updateServiceThemeWithPrevious()
			broadcastServiceUpdate()
		case MSG.NEXT_THEME:
            BPUtil.BPLog("WallpaperServicebroadcastReceiver - MSG.NEXT_THEME")
			updateServiceThemeWithNext()
			broadcastServiceUpdate()
        case MSG.TOGGLE_PAUSE:
            BPUtil.BPLog("WallpaperServicebroadcastReceiver - MSG.TOGGLE_PAUSE")
            togglePause()
		default:
            BPUtil.BPLog("WallpaperServicebroadcastReceiver -Invalid Action")
		}
	}
    
    private func onStartCommand() {
        if (!_started) {
            BPUtil.BPLog("Wallpaper Info Service Started.")
            _platformWallpaper = PlatformWallpaper()
            _imageFileManager = ImageFileManager()
            let wpsi = WallpaperServiceInfo.getInstance()
            setNewSourceRootPath(wpsi.getSourceRootPath())
			WallpaperWidgetProvider.refreshWidgets()
            _started = true;
            _platformTimer = PlatformTimer(self);
            setInterval(interval: wpsi.getInterval())
            pause_resume_service()
			// adding screen change observer
            NotificationCenter.default.addObserver(self, selector: #selector(screenChangeCallback), name: NSApplication.didChangeScreenParametersNotification, object: nil)
		} else {
			WallpaperWidgetProvider.refreshWidgets()
		}
		resetSaverTimer()
    }
	
	func resetSaverTimer() {
        BPUtil.BPLog("WallpaperService.resetSaverTimer")
		_saverTimer?.suspend()
		let wpsi = WallpaperServiceInfo.getInstance()
		if (wpsi.getSaver() && wpsi.getMode() != .wallpaper && !wpsi.getPause()) {
			_saverTimer = RepeatingTimer(timeInterval: TimeInterval(60 * wpsi.getSaverTime()))
			_saverTimer?.eventHandler = {
                BPUtil.BPLog("WallpaperService.resetSaverTimer - suspended after %d minutes of idle time.", wpsi.getSaverTime())
				self._saverTimer?.suspend()
				self._saverTimer = nil
				wpsi.setPause(true)
				self.pause_resume_service()
			}
			_saverTimer?.resume()
		} else {
			_saverTimer = nil
		}
	}
    
    @objc func screenChangeCallback() {
        BPUtil.BPLog("WallpaperService.screenChangeCallback")
        WallpaperWidgetProvider.refreshWidgets()
    }

    func wallpaperSwitchCallback() {
		// print("WallpaperService.wallpaperSwitchCallback()")
		DispatchQueue.main.async { [weak self] in
			objc_sync_enter(self!)
			let wpsi = WallpaperServiceInfo.getInstance();
			if (!wpsi.getPause()) {
				self?.naviageWallpaper(offset: 1);
			}
			objc_sync_exit(self!)
		}
    }
    
    func setInterval(interval: Int ) {
        _platformTimer?.setInterval(interval: interval)
    }
    
    private func togglePause() {
        let wpsi = WallpaperServiceInfo.getInstance()
        wpsi.setPause(!wpsi.getPause())
        pause_resume_service()
    }

    func pause_resume_service() {
        let wpsi = WallpaperServiceInfo.getInstance()
        if (wpsi.getPause()) {
            _platformTimer?.pause()
        } else {
            _platformTimer?.resume()
            DispatchQueue.main.async { [weak self] in
                self?.naviageWallpaper(offset: 1)
            }
        }
        broadcastServiceUpdate()
    }
    
    func setNewSourceRootPath(_ newSourceRootPath: String) {
        let wpsi = WallpaperServiceInfo.getInstance()
        wpsi.setcurrentWPath(nil)
        wpsi.setThumbnail(nil)
        wpsi.setSourceRootPath(newSourceRootPath)
        _imageFileManager.setSourceRootPath(newSourceRootPath)
    }
    
    func naviageWallpaper(offset: Int) {
		// wait for theme preparation is done for forward navigation
		if (!_imageFileManager.isThemeReady() && offset == 1) {
            BPUtil.BPLog("skip new image during theme preparation")
			return
		}
        let wpsi = WallpaperServiceInfo.getInstance()
		if let currentWPath = wpsi.getcurrentWPath() {
			var imagePath = _imageFileManager.retrievePathFromSource(pivotWPath: currentWPath, offset: offset)
            if (imagePath == nil) { // there is no image in source
                _platformWallpaper.makeThumbnailFromScreenWallpaper()
                broadcastServiceUpdate()
                BPUtil.BPLog("naviageWallpaper skipped by no image.")
                return;
            }

            // Check when previous image is filtered one while option is on
			while (offset == -1 && !wpsi.getThemeInfo().isThemeImage(coden: imagePath?.coden())) {
				let PreviousImagePath = _imageFileManager.retrievePathFromSource(pivotWPath: imagePath, offset: -1)
                // must avoid infinite loop by checking previous image stays same
				if (PreviousImagePath!.path == imagePath!.path) {
                    // BPUtil.BPLog("%s", "naviageWallpaper skipped by no previous unfiltered image.")
                    BPUtil.BPLog ("naviageWallpaper skipped by no previous theme image.")
                    return
				}
                imagePath = PreviousImagePath
            }

            // change Wallpaper only if there is new image
			if (imagePath!.path != currentWPath.path) {
				changeWallpaper(wpath: imagePath)
            } else {
                // BPUtil.BPLog("%s", "naviageWallpaper skipped by no previous image.")
                BPUtil.BPLog("naviageWallpaper skipped by no previous image.")
            }
        } else {
            // Change with the first Wallpaper
            if (offset == 1) {
                changeWallpaper(wpath: nil)
            } else {
                BPUtil.BPLog("naviageWallpaper skipped by no previous image.")
            }
        }
    }
    
    func setWallpaperFromLastUsedPaths(imagePath: WPath) {
        restartTimer();
        changeWallpaper(wpath: imagePath)
    }
    
    func restartTimer() {
        _platformTimer?.resetTimer()
    }
    
	private func changeWallpaper(wpath: WPath?) {
        // Only true for the first image
        let fm = FileManager.default
        var currentWPath = wpath
		if (currentWPath == nil || !(fm.fileExists(atPath: currentWPath!.path))) {
            currentWPath = _imageFileManager.retrievePathFromSource(pivotWPath: nil, offset: 1)
            if (currentWPath == nil) { // there is no image in source
                _platformWallpaper.makeThumbnailFromScreenWallpaper()
                broadcastServiceUpdate()
                BPUtil.BPLog("changeWallpaper skipped by no image.")
                return;
            }
        }

        let wpsi = WallpaperServiceInfo.getInstance()
        var count:Int = 0;
        let total:Int? = _imageFileManager.getTotalImageCount()
		while (!wpsi.getThemeInfo().isThemeImage(coden: currentWPath?.coden())) {
            currentWPath = _imageFileManager.retrievePathFromSource(pivotWPath: currentWPath, offset: 1)
            // must avoid infinite loop in case all files are filtered
            // reach end of source due to all filtered files
            count += 1
            if (count >= total!) {
                BPUtil.BPLog("changeWallpaper skipped by no theme image.")
				// pause service otherwise service is hanging
				wpsi.setPause(true)
				wpsi.setLastUsedPaths(WLinkedHashMap<String, String>())
                pause_resume_service()
                broadcastServiceUpdate()
                return
            }
        }
		let pretty_path = currentWPath!.label()
		if (_platformWallpaper.setWallpaper(path: currentWPath!.path)) {
            wpsi.setcurrentWPath(currentWPath!);
            BPUtil.BPLog("{WP} " + pretty_path + " [" + _imageFileManager.getImageStat(path:currentWPath!.path) + "]")
            broadcastServiceUpdate()
        } else {
			// code sync with windows
			wpsi.setcurrentWPath(nil);
            BPUtil.BPLog("{WP} " + pretty_path + " is failed. [" + _imageFileManager.getImageStat(path:currentWPath!.path) + "]")
        }
    }
	
	func updateServiceTheme(theme: Theme) {
		let wpsi = WallpaperServiceInfo.getInstance()
		wpsi.setThemeInfo(ThemeInfo.getThemeInfo(theme: theme))
		_imageFileManager.prepareSourceForTheme()
		WallpaperWidgetProvider.updateThemeWidget(themeLabel: theme.label)
	}
	
	func updateServiceThemeWithNext() {
		updateServiceTheme(theme: WallpaperServiceInfo.getInstance().getThemeInfo().getNextThemeInfo()._theme)
	}
	
	func updateServiceThemeWithPrevious() {
		updateServiceTheme(theme: WallpaperServiceInfo.getInstance().getThemeInfo().getPrevousThemeInfo()._theme)
	}
	
    func updateServiceCustomConfig(customConfigString: String) {
		let wpsi = WallpaperServiceInfo.getInstance()
		wpsi.setCustomThemeInfo(ThemeInfo.setCustomConfig(customConfigString: customConfigString))
		// reassign custom theme to invoke prepareTheme() with updated custom config string
		// then prepare source again with it
		if (wpsi.getTheme() == .custom) {
			wpsi.setThemeInfo(ThemeInfo.getThemeInfo(theme: .custom))
			_imageFileManager.prepareSourceForTheme()
		}
	}
	
    private func broadcastServiceUpdate() {
		//print("WallpaperService.broadcastServiceUpdate()")
		let wpsi = WallpaperServiceInfo.getInstance()
		// To Notification
		_wallpaperNotification.buildNotification(wpath: wpsi.getcurrentWPath(), thumbnail: wpsi.getThumbnail(), theme: wpsi.getTheme())
		// To Widget
		WallpaperWidgetProvider.updatePauseWidget(pause:wpsi.getPause())
		WallpaperWidgetProvider.updateThemeWidget(themeLabel: wpsi.getTheme().label, forceShow: wpsi.getPause())
		if let currentWPath = wpsi.getcurrentWPath() {
			WallpaperWidgetProvider.updateLabelWidget(text: currentWPath.label())
		} else {
            WallpaperWidgetProvider.updateLabelWidget(text: "")
		}
		// To UI
		_wallpaperServiceHandler.replyToClient(thumbnail: wpsi.getThumbnail(), currentWPath: wpsi.getcurrentWPath(), pause: wpsi.getPause())
    }
    
    public func activateHotkey() {
        _wallpaperNotification._platformHotKey.initEventPort()
    }
    
    public func runBPHTTPHandler() {
        let group = MultiThreadedEventLoopGroup(numberOfThreads: ProcessInfo.processInfo.activeProcessorCount)
        let bootstrap = ServerBootstrap(group: group)
            .childChannelInitializer { channel in
                channel.pipeline.configureHTTPServerPipeline().flatMap {
                    channel.pipeline.addHandler(BPHTTPHandler())
                }
            }
            .bind(host: "localhost", port: 8080)
        let channel = try? bootstrap.wait()
        if (channel != nil) {
            // BPUtil.BPLog("Server running on \(channel?.localAddress!)")
            try? channel?.closeFuture.wait()
            // running
        }
    }
	
    private var _imageFileManager: ImageFileManager = ImageFileManager()
    private var _platformTimer: PlatformTimer? = nil
	private var _saverTimer:RepeatingTimer? = nil
    private var _started : Bool = false
    lazy private var _platformWallpaper: PlatformWallpaper = PlatformWallpaper()
	lazy public var _wallpaperServiceHandler: WallpaperServiceHandler = WallpaperServiceHandler()

    // service const values
    public static var _maxLastUsedPaths : Int = 50
	
	// only for Mac OS X, Windows, Linux
	private var _wallpaperNotification: WallpaperNotification = WallpaperNotification()
    func isStarted() -> Bool { return _started }
}

File: apps/mobile-android/wallpaperinfo/app/src/main/java/com/thirdwavesoft/wallpaperinfo/WallpaperService.java
package com.thirdwavesoft.wallpaperinfo;

import static android.content.Intent.ACTION_CONFIGURATION_CHANGED;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.res.Configuration;
import android.os.IBinder;
import android.util.Log;

import androidx.annotation.Nullable;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

public class WallpaperService extends Service {

    @Override
    public void onCreate() {
    	Log.d(_TAG, "onCreate()");
		WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
		wpsi.setAppContext(this);
		BPUtil.BPLog("%s", "Model : " + android.os.Build.MODEL);
        IntentFilter filter = new IntentFilter();
        filter.addAction(ACTION_CONFIGURATION_CHANGED);
        filter.addAction("ACTION_SERVICE_REQUEST_INFO");
        filter.addAction("ACTION_SERVICE_IMAGE_PAUSE_RESUME");
        filter.addAction("ACTION_SERVICE_IMAGE_NEXT");
        filter.addAction("ACTION_SERVICE_IMAGE_PREVIOUS");
        registerReceiver(_broadcastReceiver, filter);
		_wallpaperServiceHandler = new WallpaperServiceHandler(this);
    }

	@Override
	public void onDestroy() {
		_imageFileManager.stopWatching();
    	// stop timer when service is destroyed to prevent service is rerunning
		_platformTimer.pause();
		// update widget for default view
		WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
		wpsi.setcurrentWPath(null);
		wpsi.setThumbnail(null);
		broadcastServiceUpdate();
		// unregister broadcast
		unregisterReceiver(_broadcastReceiver);
		_started = false;
		super.onDestroy();
	}

	public BroadcastReceiver _broadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent myIntent) {
        	String action = myIntent.getAction();
			Log.i(_TAG, "Broadcast Received in Service : " + action);
			switch(action) {
				case ACTION_CONFIGURATION_CHANGED:
					orientationUpdate();
					break;
				case "ACTION_SERVICE_REQUEST_INFO":
					broadcastServiceUpdate();
					break;
				case "ACTION_SERVICE_IMAGE_PREVIOUS":
					BPUtil.BPLog("%s", "MSG_PREVIOUS: ACTION_SERVICE_IMAGE_PREVIOUS");
					restartTimer();
					naviageWallpaper(-1);
					break;
				case "ACTION_SERVICE_IMAGE_NEXT":
					BPUtil.BPLog("%s", "MSG_NEXT: ACTION_SERVICE_IMAGE_NEXT");
					restartTimer();
					naviageWallpaper(1);
					break;
				case "ACTION_SERVICE_IMAGE_PAUSE_RESUME":
					BPUtil.BPLog("%s", "MSG_TOGGLE_PAUSE: ACTION_SERVICE_IMAGE_PAUSE_RESUME");
					togglePause();
					break;
				default:
					return;
			}
			resetSaverTimer();
        }
    };

	@Override
	public int onStartCommand(Intent intent, int flags, int startId) {
		super.onStartCommand(intent, flags, startId);
		if (!_started) {
			BPUtil.BPLog("%s", "Wallpaper Info Service Started.");
			Notification notification = WallpaperNotification.buildNotification(this, null, null);
			startForeground(_notificationId, notification);
			_platformWallpaper = new PlatformWallpaper(this);
			_imageFileManager = new ImageFileManager();
			WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
			setNewSourceRootPath(wpsi.getSourceRootPath());
			_started = true;
			_platformTimer = new PlatformTimer(this);
			setInterval(wpsi.getInterval());
			pause_resume_service();
		} else {
			orientationUpdate();
		}
		resetSaverTimer();
		return START_STICKY;
    }

    synchronized public void resetSaverTimer() {
		Log.i(_TAG, "resetSaverTimer - reset saver timer.");
		if (_saverTimer != null) {
			_saverTimer.cancel(true);
		}
		final Runnable saverTimerFired = new Runnable() {
			public void run() {
			WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
			Log.i(_TAG, "resetSaverTimer - suspend service after " + wpsi.getSaverTime() +" minutes of idle time.");
			_saverTimer.cancel(true);
			_saverTimer = null;
			wpsi.setPause(true);
			pause_resume_service();
			broadcastServiceUpdate();
			}
		};
		WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
		if (wpsi.getSaver() && wpsi.getMode() != ServiceMode.wallpaper && !wpsi.getPause()) {
			_saverTimer =  _scheduler.scheduleWithFixedDelay(saverTimerFired, wpsi.getSaverTime() * 60, wpsi.getSaverTime() * 60, TimeUnit.SECONDS);
		} else {
			_saverTimer = null;
		}
	}
	
	synchronized public void wallpaperSwitchCallback() {
		if (PlatformInfo.isScreenOn(this)) {
			_changedAsOffLastTime = false;
			if (PlatformInfo.isHomeScreenTop(this)) {
				_changedAsHiddenLastTime = false;
			} else if (!_changedAsHiddenLastTime) {
				_changedAsHiddenLastTime = true;
			} else {
				return;
			}
		} else {
			_changedAsHiddenLastTime = false;
			if (_changedAsOffLastTime) {
				return;
			}
			_changedAsOffLastTime = true;
		}
		naviageWallpaper(1);
	}

	public void setInterval(int interval) {
		_platformTimer.setInterval(interval);
	}

	private void togglePause() {
		WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
		wpsi.setPause(!wpsi.getPause());
		pause_resume_service();
	}
	
	public void pause_resume_service() {
		WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
		if (wpsi.getPause()) {
			_platformTimer.pause();
		} else {
			_platformTimer.resume();
			naviageWallpaper(1);
		}
		broadcastServiceUpdate();
	}

	synchronized public void setNewSourceRootPath(String newSourceRootPath) {
		WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
		wpsi.setcurrentWPath(null);
		wpsi.setThumbnail(null);
		wpsi.setSourceRootPath(newSourceRootPath);
		_imageFileManager.setSourceRootPath(newSourceRootPath);
	}

	synchronized public void naviageWallpaper(int offset) {
		// wait for theme preparation is done for forward navigation
		if (!_imageFileManager.isThemeReady() && offset == 1) {
			BPUtil.BPLog("%s", "skip new image during theme preparation");
			return;
		}
		WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
		WPath currentWPath = wpsi.getcurrentWPath();
		if (currentWPath != null) {
			WPath imagePath = _imageFileManager.retrievePathFromSource(currentWPath, offset);
			if (imagePath == null) { // there is no image in source
				_platformWallpaper.makeThumbnailFromScreenWallpaper();
				broadcastServiceUpdate();
				BPUtil.BPLog("%", "naviageWallpaper skipped by no image.");
				return;
			}
			// Check when previous image is filtered one while option is on
			while (offset == -1 && !wpsi.getThemeInfo().isThemeImage(imagePath.coden())) {
				WPath PreviousImagePath = _imageFileManager.retrievePathFromSource(imagePath, -1);
				// must avoid infinite loop by checking previous image stays same
				if (PreviousImagePath.path.equals(imagePath.path)) {
					BPUtil.BPLog("%s", "naviageWallpaper skipped by no previous unfiltered image.");
					return;
				}
				imagePath = PreviousImagePath;
			}
			// same path check removed here and move to where setting actual wallpaper due to mode introduction
			changeWallpaper(imagePath);
		} else {
			// Change with the first Wallpaper
			if (offset == 1) {
				changeWallpaper(null);
			} else {
				BPUtil.BPLog("%s", "naviageWallpaper skipped by no previous image.");
			}
		}
	}

	synchronized public void setWallpaperFromLastUsedPaths(WPath imagePath) {
		restartTimer();
		changeWallpaper(imagePath);
	}

	synchronized public void restartTimer() {
		_platformTimer.resetTimer();
	}

	synchronized private void changeWallpaper(WPath currentWPath) {
		// Only true for the first image
		if (currentWPath == null || !BPUtil.fileExists(currentWPath.path)) {
			currentWPath = _imageFileManager.retrievePathFromSource(null, 1);
			if (currentWPath == null) { // there is no image in source
				_platformWallpaper.makeThumbnailFromScreenWallpaper();
				broadcastServiceUpdate();
				BPUtil.BPLog("%s", "changeWallpaper skipped by no image.");
				return;
			}
		}
		WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
		int count = 0, total = _imageFileManager.getTotalImageCount();
		while (!wpsi.getThemeInfo().isThemeImage(currentWPath.coden())) {
			currentWPath = _imageFileManager.retrievePathFromSource(currentWPath, 1);
			// must avoid infinite loop in case all files are filtered
			// reach end of source due to all filtered files
			if (count++ >= total) {
				BPUtil.BPLog("%s", "changeWallpaper skipped by theme image.");
				// pause service otherwise service is hanging
				wpsi.setPause(true);
				pause_resume_service();
				wpsi.getLastUsedPaths().clear();
				broadcastServiceUpdate();
				return;
			}
		}
		if (_platformWallpaper.setWallpaper(currentWPath)) {
			wpsi.setcurrentWPath(currentWPath);
			BPUtil.BPLog("%s", "{WP} " + currentWPath.label() + " [" + _imageFileManager.getImageStat(currentWPath.path) + "]");
			broadcastServiceUpdate();
		} else {
			// code sync with windows
			wpsi.setcurrentWPath(null);
			BPUtil.BPLog("%s", "{WP} " + currentWPath.label() + " is failed. [" + _imageFileManager.getImageStat(currentWPath.path) + "]");
		}
	}
	
	void updateServiceTheme(Theme theme) {
		WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
		wpsi.setThemeInfo(ThemeInfo.getThemeInfo(theme));
		_imageFileManager.prepareSourceForTheme();
		broadcastServiceUpdate();
	}

	void updateServiceThemeWithNext() {
		updateServiceTheme(WallpaperServiceInfo.getInstance().getThemeInfo().getNextThemeInfo()._theme);
	}

	void updateServiceThemeWithPrevious() {
		updateServiceTheme(WallpaperServiceInfo.getInstance().getThemeInfo().getPrevousThemeInfo()._theme);
	}

    void updateServiceCustomConfig(String customConfigString) {
		WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
		wpsi.setCustomThemeInfo(ThemeInfo.setCustomConfig(customConfigString));
		// reassign custom theme to invoke prepareTheme() with updated custom config string
		// then prepare source again with it
		if (wpsi.getTheme() == Theme.custom) {
			wpsi.setThemeInfo(ThemeInfo.getThemeInfo(Theme.custom));
			_imageFileManager.prepareSourceForTheme();
		}
	}

	void orientationUpdate() {
		if(getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE) {
			BPUtil.BPLog("%s", "LANDSCAPE");
			_platformWallpaper.orientationUpdate(Configuration.ORIENTATION_LANDSCAPE);
		} else {
			BPUtil.BPLog("%s", "PORTRAIT");
			_platformWallpaper.orientationUpdate(Configuration.ORIENTATION_PORTRAIT);
		}
	}

	synchronized private void broadcastServiceUpdate() {
		Log.d(_TAG, "WallpaperService.broadcastServiceUpdate()");
		// Notification
		WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
		Notification notification = WallpaperNotification.buildNotification(this, wpsi.getcurrentWPath(), wpsi.getThumbnail());
		NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
		notificationManager.notify(_notificationId, notification);
		// WallpaperInfoApp -> ACTION_GET_INFO_INTERNAL by LocalBroadcastManager
		Intent intent = new Intent("ACTION_GET_INFO_FROM_SERVICE");
		intent.putExtra("INTENT_EXTRA_THUMBNAIL", wpsi.getThumbnail());
		intent.putExtra("INTENT_EXTRA_THEME", wpsi.getTheme().intValue());
		WPath wpath = wpsi.getcurrentWPath();
		if (wpath != null) {
			intent.putExtra("INTENT_EXTRA_IMAGE_PATH", wpath.path);
			intent.putExtra("INTENT_EXTRA_IMAGE_EXIF", wpath.exif);
		} else {
			intent.putExtra("INTENT_EXTRA_IMAGE_PATH", (String)null);
			intent.putExtra("INTENT_EXTRA_IMAGE_EXIF", "");
		}
		intent.putExtra("INTENT_EXTRA_PAUSE", WallpaperServiceInfo.getInstance().getPause());
		sendBroadcast(intent);
		// Widget
		WallpaperWidgetProvider.updateWidget(this);
	}

	@Nullable
	@Override
	public IBinder onBind(Intent intent) {
		Log.d(_TAG, "onBind()");
		return _wallpaperServiceHandler.getBinder();
	}

	private ImageFileManager _imageFileManager = null;
	private PlatformTimer _platformTimer = null;
	private ScheduledFuture<?> _saverTimer = null;
	private boolean _started = false;
	private PlatformWallpaper _platformWallpaper;
	private WallpaperServiceHandler _wallpaperServiceHandler;


	// service const values
	public final static int _maxLastUsedPaths = 50;

	// only for Android
	private boolean _changedAsHiddenLastTime = false;
	private boolean _changedAsOffLastTime = false;
	private final int _notificationId = 13;
	private final ScheduledExecutorService _scheduler =	Executors.newScheduledThreadPool(1);
	private static final String _TAG = "WallpaperService";
}
